### PR TITLE
Do not order all Yaml maps by keys, only the Helm values file

### PR DESCRIPTION
**Pull request by Jurrie**
Signed-off-by: Jurrie Overgoor <eclipse.org@jurr.org>## Description

Fixes # (issue)

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [x] My code follows the style guidelines of this project
 - [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [ ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [x] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [😀] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [x] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2689 (_2024-02-16T22:13:50Z_)

:gear: [JKube E2E Tests (7936836215)](https://github.com/jkubeio/ci/actions/runs/7936836215) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2689)](https://github.com/jkubeio/ci/actions/runs/7936836215/job/21672924045)

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/Serialization.java [hunk 1]
```diff
@@ -42,7 +46,6 @@ public class Serialization {
   static {
     for (ObjectMapper mapper : new ObjectMapper[]{JSON_MAPPER, YAML_MAPPER}) {
       mapper.enable(SerializationFeature.INDENT_OUTPUT)
-        .enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
```
Comment (Jurrie) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/Serialization.java:45]: I need to keep the ordering as-is, so we do not reorder Helm lines.

File: jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/auth/AuthConfigTest.java [hunk 2]
```diff
@@ -36,7 +36,7 @@ void simpleConstructor() {
     // Since Base64.decodeBase64 handles URL-safe encoding, must explicitly check
     // the correct characters are used
     assertThat(config.toHeaderValue(new KitLogger.SilentLogger()))
-      .isEqualTo("eyJlbWFpbCI6InJvbGFuZEBqb2xva2lhLm9yZyIsInBhc3N3b3JkIjoiIz5zZWNyZXRzPz8iLCJ1c2VybmFtZSI6InJvbGFuZCJ9");
+      .isEqualTo("eyJ1c2VybmFtZSI6InJvbGFuZCIsInBhc3N3b3JkIjoiIz5zZWNyZXRzPz8iLCJlbWFpbCI6InJvbGFuZEBqb2xva2lhLm9yZyJ9");
```
Comment (Jurrie) [inline jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/auth/AuthConfigTest.java:39]: This was changed because of https://github.com/eclipse/jkube/pull/2689#discussion_r1493034414.
Previously it read `{"email":"roland@jolokia.org","password":"#>secrets??","username":"roland"}` and now it reads `{"username":"roland","password":"#>secrets??","email":"roland@jolokia.org"}`

File: jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/auth/RegistryAuthTest.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ private void check(RegistryAuth config) {
         // Since Base64.decodeBase64 handles URL-safe encoding, must explicitly check
         // the correct characters are used
         assertThat(config.toHeaderValue())
-            .isEqualTo("eyJlbWFpbCI6InJvbGFuZEBqb2xva2lhLm9yZyIsInBhc3N3b3JkIjoiIz5zZWNyZXRzPz8iLCJ1c2VybmFtZSI6InJvbGFuZCJ9");
+            .isEqualTo("eyJ1c2VybmFtZSI6InJvbGFuZCIsInBhc3N3b3JkIjoiIz5zZWNyZXRzPz8iLCJlbWFpbCI6InJvbGFuZEBqb2xva2lhLm9yZyJ9");
```
Comment (Jurrie) [inline jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/auth/RegistryAuthTest.java:53]: This was changed because of https://github.com/eclipse/jkube/pull/2689#discussion_r1493034414.
Previously it read `{"email":"roland@jolokia.org","password":"#>secrets??","username":"roland"}` and now it reads `{"username":"roland","password":"#>secrets??","email":"roland@jolokia.org"}`

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/TemplateUtil.java [hunk 4]
```diff
@@ -13,49 +13,152 @@
...
+
 public class TemplateUtil {
+  private static final String HELM_DIRECTIVE_REGEX = "\\{\\{.*\\}\\}";
 
   private TemplateUtil() {
   }
 
   /**
-   * Ported from https://github.com/fabric8io/fabric8-maven-plugin/commit/d6bdaa37e06863677bc01cefa31f7d23c6d5f0f9
+   * This function will replace all Helm directives with a valid Yaml line containing the base64 encoded Helm directive.
```
Comment (Jurrie) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/TemplateUtil.java:26]: image

**Comment by manusa** [comment]
Didn't give a thorough look, but it seems that with this PR we could finally mix in invalid Helm/Go template syntax withing the JKube-processed Kubernetes resources (those stored at src/main/jkube).

I'm not sure if we're considering here corner cases such as `{{ include "}}something{{}}}}{{" . }}`.
Relying on regex for this kind of processing is always dangerous, so considering these scenarios in a specific test-suite is needed.

Regarding the PR, I think it probably needs proper breaking down so that it's easier to evaluate each of the things addressed here:
- Serialiation/Deserialization of common-resources and Helm
- values.yaml sorting without modifying the global Jackson configuration
- Inclusion of Helm syntax on regular yaml fragments

**Comment by Jurrie** [comment]
Hi @manusa, thanks for your reply.

> Didn't give a thorough look, but it seems that with this PR we could finally mix in invalid Helm/Go template syntax withing the JKube-processed Kubernetes resources (those stored at src/main/jkube).

That's the intention at least 😉

> I'm not sure if we're considering here corner cases such as `{{ include "}}something{{}}}}{{" . }}`. Relying on regex for this kind of processing is always dangerous, so considering these scenarios in a specific test-suite is needed.

I agree on the regex part. Also, in my opinion regex's are less maintainable than regular code. I'll see what I can do to convert this to use regular string parsing.

The corner case that you mention - is that even valid? I'll try to include such a corner case in the unit test.

> Regarding the PR, I think it probably needs proper breaking down so that it's easier to evaluate each of the things addressed here:
> 
>     * Serialiation/Deserialization of common-resources and Helm
> 
>     * values.yaml sorting without modifying the global Jackson configuration
> 
>     * Inclusion of Helm syntax on regular yaml fragments

Fine by me to break this down. The part "values.yaml sorting without modifying the global Jackson configuration" is already an available PR; please see https://github.com/eclipse/jkube/pull/2702.

As for the other two parts: I'm not sure if they stand on their own. In my opinion that should be a single PR. If you disagree, could you explain how you would like to see those two PRs?

Also, I noted in the description:
> The Helm directives will be kept in the Yaml fragments. This breaks the inner-loop development cycle. (I have thought of a solution to this, but I'd like to discuss that first, and I am hoping that implementation can be done in a separate PR.)

Do you feel comfortable to merge this feature anyway?

**Comment by manusa** [comment]
> As for the other two parts: I'm not sure if they stand on their own. In my opinion that should be a single PR. If you disagree, could you explain how you would like to see those two PRs?

Once this PR is rebased I'll probably have a clear view. During my initial review, I saw you changed and added methods to the generic Serialization tools, this is what I properly need to understand.

> > The Helm directives will be kept in the Yaml fragments. This breaks the inner-loop development cycle. (I have thought of a solution to this, but I'd like to discuss that first, and I am hoping that implementation can be done in a separate PR.)
>
> Do you feel comfortable to merge this feature anyway?

Haven't checked your implementation yet, but the idea for the inner-loop (k8s:resource) would be to ignore the helm-specific sections and warn the user about it.

**Comment by Jurrie** [comment]
> > As for the other two parts: I'm not sure if they stand on their own. In my opinion that should be a single PR. If you disagree, could you explain how you would like to see those two PRs?
> 
> Once this PR is rebased I'll probably have a clear view. During my initial review, I saw you changed and added methods to the generic Serialization tools, this is what I properly need to understand.

The PR is rebased. The basic idea behind the changes in Serialization is to reroute all methods that have different type inputs to a single method that does marshalling / unmarshalling. That's the point where I do the "mask the Helm directives" / "unmask the Helm directives".

> > > The Helm directives will be kept in the Yaml fragments. This breaks the inner-loop development cycle. (I have thought of a solution to this, but I'd like to discuss that first, and I am hoping that implementation can be done in a separate PR.)
> > 
> > 
> > Do you feel comfortable to merge this feature anyway?
> 
> Haven't checked your implementation yet, but the idea for the inner-loop (k8s:resource) would be to ignore the helm-specific sections and warn the user about it.

That is not what is currently happening. The Helm directives will still be there in the final output. `k8s:resource` will keep them in. I don't see a way around this, as Helm directives can (and probably will) change the way that the Yaml files are applied to Kubernetes / OpenShift. If you would ignore them, you would end up with totally different behavior for the inner-loop and the outer-loop.

I do have a solution to this though. Maybe it's best to first introduce the goals `helm-install` and `helm-uninstall`. When a user has Helm directives in the Yaml files, the user would need to use `k8s:helm-install` instead of `k8s:deploy` and `k8s:helm-uninstall` instead of `k8s:undeploy`.
